lib/puppet/provider/puppet_certificate/ruby.rb

```diff
@@ -142,109 +203,50 @@ def dns_alt_names=(should_dns_alt_names)
+      debug("Attempting to load key for #{@resource[:name]}")
+      @key ||= @cert_provider.load_private_key(@resource[:name])
+    else
+      @key ||= Puppet::SSL::Key.indirection.find(@resource[:name])
+    end
   end
 
   def certificate
-    @certificate ||= Puppet::SSL::Certificate.indirection.find(@resource[:name])
-  end
-
-  def csr
-    @csr ||= begin
-      @csr = Puppet::SSL::CertificateRequest.indirection.find(@resource[:name])
-      unless @csr
-        @csr = Puppet::SSL::CertificateRequest.new(@resource[:name])
-        @csr.generate(key.content, options)
+    if at_least_puppet6
+      debug("Attempting to load cert for #{@resource[:name]}") unless @certificate
+      @certificate ||= @cert_provider.load_client_cert(@resource[:name])
+    else
+      if (cert = Puppet::SSL::Certificate.indirection.find(@resource[:name]))
+        @certificate ||= cert.content
       end
-      @csr
     end
   end
 
   def is_valid?
-      unless certificate.nil?
-          grace_time = @resource[:renewal_grace_period] * 60 * 60 * 24
-          certificate.content.not_after - grace_time > Time.now
-      end
-  end
+    return if certificate.nil?
 
-  # Create/return a store that uses our SSL info to validate
-  # connections.
-  def ssl_store(purpose = OpenSSL::X509::PURPOSE_ANY)
-    if @ssl_store.nil?
-      @ssl_store = build_ssl_store(purpose)
-    end
-    @ssl_store
+    grace_time = @resource[:renewal_grace_period] * 60 * 60 * 24
+    certificate.not_after - grace_time > Time.now
   end
 
-  def build_ssl_store(purpose=OpenSSL::X509::PURPOSE_ANY)
-    store = OpenSSL::X509::Store.new
-    store.purpose = purpose
-
-    # Use the file path here, because we don't want to cause
-    # a lookup in the middle of setting our ssl connection.
-    store.add_file(Puppet.settings[:localcacert])
-
-    if use_crl?
-      if !Puppet::FileSystem.exist?(crl_path)
-        download_and_save_crl_bundle(store)
-      end
-
-      crls = load_crls(crl_path)
-
-      flags = OpenSSL::X509::V_FLAG_CRL_CHECK
-      if use_crl_chain?
-        flags |= OpenSSL::X509::V_FLAG_CRL_CHECK_ALL
-      end
-
-      store.flags = flags
-      crls.each {|crl| store.add_crl(crl) }
-    end
-    store
+  def debug(msg)
+    Puppet.debug "puppet_certificate: #{msg}"
   end
 
-  def use_crl?
-    !!@crl_usage
+  def warning(msg)
+    Puppet.warning "puppet_certificate: #{msg}"
   end
 
-  def use_crl_chain?
-    @crl_usage == true || @crl_usage == :chain
+  def fingerprint(cert)
+    Puppet::SSL::Digest.new(nil, cert.to_der)
   end
 
-  def get_certificate(certname)
-    Puppet::Rest::Routes.get_certificate(
-      certname,
-      Puppet::SSL::SSLContext.new(store: ssl_store)
-    )
+  def create_route(ssl_context)
+    Puppet.runtime[:http].create_session.route_to(:ca, ssl_context: ssl_context)
   end
 
-  def debug(msg)
-    Puppet.debug "puppet_certificate: #{msg}"
+  def at_least_puppet6
+    Gem::Version.new(Puppet.version) > Gem::Version.new('6.0.0')
```

Comment:
```suggestion
    Gem::Version.new(Puppet.version) >= Gem::Version.new('6.0.0')
```

Reply:
In reality, I think 6.x less that 6.4 are likely broken anyway.  I don't suppose that's a problem, but I'll add some fail message for this.

Reply:
Fixed